// http-session/core/Hsess.js
/**
*  Copyright (c) 2016 Abdennour TOUMI <http://abdennoor.com> ceo@rathath-it.com

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*
**/
(() => {
    class _ extends Map {

        constructor(id) {
            super();
            id = id || new Date().getTime() + '';
            this.___settings___ = {
                id: id,
                history: [],
                valid:true
            };
        }

        getId() {
            return this.___settings___.id;
        }




        getAttribute(k) {
            return this[k];
        }

        getAttributeNames() {
            return Object.keys(this).filter(k => k !== '___settings___');
        }

        setAttribute(k, v) {

            this[k] = v;
        }

        setMaxInactiveInterval(seconds) {
          this.___settings___.maxInactiveInterval=seconds;
        }





        get creationTime() {
            if (this.getId().length > 20) {
                return new Date(parseInt(this.getId().substr(20)));
            }
            return new Date(parseInt(this.getId()));
        }
        get lastAccess() {
            if (this.___settings___.history.length) {
                return this.___settings___.history[this.___settings___.history.length - 1].at;
            }
            return this.creationTime;
        }
        addHistroy(rq) {
            this.___settings___.history.push({
                at: new Date(),
                request: rq
            });
        }

    }

    module.exports = _;
})()
